### DISCUSSION TITLE
Why InPort.registerConsumerEvent don't establish a precedence between InPort >> Consumer ?

### DISCUSSION THREAD

**Discussion by myhwz** [sparcians/map#504] [General]
hi,  viewing the code from sparta/ports/port.cpp:
1) from the OutPort.bind(...), there is a comment:
            //          ,--------------------------------------------------.
            //          |                                                                         V
            // producer -> [      inport internal delivery      ]* -> consumer
     The inport -> consumer precedence is established during the consumer registration from a call to: 
     InPort::registerConsumerEvent()

2) but in Inport.registerConsumerEvent(...): 
    I can't find the code to establish the precedence of:  InPort >> Consumer.    
    What's the considerations in here ?

    thanks a lot.

**Comment by klingaard** [comment]
The call to `registerConsumerEvent()` places the given event into a list called `port_consumers_` on the `InPort`: https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/Port.hpp#L378

That list is used in binding here:
https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/Port.hpp#L619

The actual precedence is made during the call `setProducerPrecedence_` here:
https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/Port.hpp#L630

and eventually established to the internal event (Payload for example) here:
https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/DataPort.hpp#L458

Hope this clarifies how this works.

**Comment by myhwz** [comment]
Sorry to ask the question in the improper place, and thanks for you answer.
In OutPort.bind(...):
```c++
if((inp->getPortDelay() == 0) && presume_zero_delay_) {
    for(auto & pd : port_producers_)
    {
        for(auto & cons : inp->port_consumers_) {
            if(pd->getSchedulingPhase() == cons->getSchedulingPhase()) {
                pd->precedes(cons, reason);

                // Here just add a precedence between:  Producer >> Consumers
                // the Consumers are saved in: registerConsumerEvent()
            }
        }
        inp->setProducerPrecedence_(pd);

        // Here just add a precedence between:  Producer >> Inport.InnerEvent
    }
}
```
So after Bind,  the DAG is:
```mermaid
graph LR
1(Producer) --> 2(Consumers)
1 --> 3("InPort's InnerEvent")
3 -.-x 2
```

My question is:  
**Why not establish a precedence between:  `InPort's InnerEvent  >> Consumers`  ?**

thanks a lot.

**Reply by klingaard** [reply]
Good question!  And you had me stumped for a minute.  :laughing:  I thought that this was possibly an oversight until I realized I forgot tell you about the last bit of binding activity that occurs in the port.

After the port consumers are "connected" to the port producers (mostly for, and only needed for, 0-cycle delay ports), a call to the InPort's internal `bind_` is called;

https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/Port.hpp#L635

And in _that function_ the inner event is preceded to the consumers:

https://github.com/sparcians/map/blob/3092846b5e4d7ef461d173af10c1ad4a47f7166f/sparta/sparta/ports/DataPort.hpp#L478-L484
